lib/streamlit/elements/widgets/data_editor.py

```diff
@@ -288,7 +301,7 @@ def _apply_cell_edits(
                 )
             else:
                 col_pos = df.columns.get_loc(col_name)
-                df.iloc[row_pos, col_pos] = _parse_value(
+                df.iat[row_pos, col_pos] = _parse_value(
```

Comment:
We need to switch from `iloc` to `iat` here, since `iloc` interprets list values differently. But `iat` is also the more efficient solution in general to overwrite single values.